tests/integration/cag/test_range.py

```diff
@@ -0,0 +1,290 @@
+
+    # Assert
+    expected_updated_metadata = Metadata.load_from_dict({
+        'columns': {
+            'A': {'sdtype': 'numerical'},
+            'A#B': {'sdtype': 'numerical'},
+            'B#C': {'sdtype': 'numerical'},
+        }
+    }).to_dict()
+    assert expected_updated_metadata == updated_metadata.to_dict()
+    assert list(transformed.columns) == ['A', 'A#B', 'B#C', 'A#B#C.nan_component']
+    pd.testing.assert_frame_equal(
+        pd.concat([data.iloc[:2], data.iloc[3:]]),
+        pd.concat([reverse_transformed.iloc[:2], reverse_transformed.iloc[3:]]),
+    )
+    assert 1 < reverse_transformed.iloc[2]['B'] < 30
+    assert pd.isna(reverse_transformed.iloc[2]['A'])
+    assert 100 < reverse_transformed.iloc[2]['C'] < 300
+
+
+def test_range_pattern_datetime():
+    """Test that Range pattern works with datetime columns."""
+    # Setup
+    data = pd.DataFrame({
+        'A': [
+            pd.Timestamp('2024-01-01'),
+            pd.Timestamp('2024-01-02'),
+            pd.Timestamp('2024-01-03'),
+            pd.Timestamp('2024-01-01'),
+            pd.Timestamp('2024-01-02'),
+            pd.Timestamp('2024-01-01'),
+        ],
+        'B': [
+            pd.Timestamp('2024-01-02'),
+            pd.Timestamp('2024-01-03'),
+            pd.Timestamp('2024-01-04'),
+            pd.Timestamp('2024-01-05'),
+            pd.Timestamp('2024-01-06'),
+            pd.Timestamp('2024-01-07'),
+        ],
+        'C': [
+            pd.Timestamp('2024-01-03'),
+            pd.Timestamp('2024-01-04'),
+            pd.Timestamp('2024-01-05'),
+            pd.Timestamp('2024-01-06'),
+            pd.Timestamp('2024-01-07'),
+            pd.Timestamp('2024-01-08'),
+        ],
+    })
+    metadata = Metadata.load_from_dict({
+        'columns': {
+            'A': {'sdtype': 'datetime'},
+            'B': {'sdtype': 'datetime'},
+            'C': {'sdtype': 'datetime'},
+        }
+    })
+    pattern = Range(
+        low_column_name='A',
+        middle_column_name='B',
+        high_column_name='C',
+    )
+
+    # Run
+    updated_metadata, transformed, reverse_transformed = run_pattern(pattern, data, metadata)
+
+    # Assert
+    expected_updated_metadata = Metadata.load_from_dict({
+        'columns': {
+            'A': {'sdtype': 'datetime'},
+            'A#B': {'sdtype': 'numerical'},
+            'B#C': {'sdtype': 'numerical'},
+        }
+    }).to_dict()
+    assert expected_updated_metadata == updated_metadata.to_dict()
+    assert list(transformed.columns) == ['A', 'A#B', 'B#C']
+
+    # Check that the timestamps are very close to each other
+    for col in ['A', 'B', 'C']:
+        diff = (data[col] - reverse_transformed[col]).abs()
+        assert (diff.dt.total_seconds() < 1e-6).all()
+
+
+def test_range_pattern_datetime_nans():
+    """Test that Range pattern works with datetime columns with NaNs."""
+    # Setup
+    data = pd.DataFrame({
+        'A': [
+            np.nan,
```

Comment:
For datetimes columns, I believe there are 2 possibilities. A `datetime64[ns]` dtype (numpy) and an `object` dtype. 
We should also test `object` dtype: 
```python
pd.Series([pd.Timestamp('2024-01-02'), np.nan], dtype="object")
pd.Series([pd.Timestamp('2024-01-02'), pd.NA], dtype="object")
```

Reply:
The `validate` method crashes if the datetimes are passed as strings, since there is no logic to convert them to numerical values.

As I understand it, the goal of this issue is to convert the existing constraint into a CAG. Any additional changes should probably be done in a separate PR. We had a similar [discussion](https://github.com/sdv-dev/SDV/pull/2405#discussion_r1983648731) on the `Inequality` constraint.

It is a good point though @gsheni, I agree this should be addressed.

Reply:
Gotcha. Let's maintain existing functionality then